// @keystone-alpha/fields/dist/chunk-b35144d1.cjs.js
'use strict';

function _interopDefault(ex) {
  return ex && typeof ex === 'object' && 'default' in ex ? ex['default'] : ex;
}

var _objectWithoutProperties = _interopDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _defineProperty = _interopDefault(require("@babel/runtime/helpers/defineProperty"));

var core = require('@emotion/core');

var React = require('react');

var React__default = _interopDefault(React);

var fields = require('@arch-ui/fields');

var _extends = _interopDefault(require("@babel/runtime/helpers/extends"));

var theme = require('@arch-ui/theme');

var input = require('@arch-ui/input');

var __chunk_3 = require('./chunk-2e9639ec.cjs.js');

const StyledPreview = _ref => {
  let preview = _ref.preview,
      originalUrl = _ref.originalUrl,
      fieldPath = _ref.fieldPath,
      props = _objectWithoutProperties(_ref, ["preview", "originalUrl", "fieldPath"]);

  return core.jsx(__chunk_3.PreviewPlaceholder, _extends({
    data: preview,
    originalUrl: originalUrl,
    fieldPath: fieldPath,
    css: {
      backgroundColor: 'white',
      borderRadius: theme.borderRadius,
      border: "1px solid ".concat(theme.colors.N20),
      marginTop: theme.gridSize,
      padding: 4,
      width: 410,
      // 300px image + chrome
      boxSizing: 'border-box'
    }
  }, props));
};

const PlaceholderPreview = _ref2 => {
  let originalUrl = _ref2.originalUrl,
      fieldPath = _ref2.fieldPath;
  return core.jsx(StyledPreview, {
    data: {
      html: '<div style="background-color: darkgray; height: 2.5em" />',
      title: 'Preview will be generated after save'
    },
    originalUrl: originalUrl,
    fieldPath: fieldPath,
    css: {
      opacity: 0.3
    }
  });
};

class OEmbedField extends React.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "onChange", event => {
      this.props.onChange({
        originalUrl: event.target.value
      });
    });
  }

  render() {
    const _this$props = this.props,
          autoFocus = _this$props.autoFocus,
          field = _this$props.field,
          _this$props$value = _this$props.value,
          value = _this$props$value === void 0 ? null : _this$props$value,
          _this$props$savedValu = _this$props.savedValue,
          savedValue = _this$props$savedValu === void 0 ? null : _this$props$savedValu,
          errors = _this$props.errors;
    const htmlID = "ks-oembed-".concat(field.path);
    const canRead = errors.every(error => !(error instanceof Error && error.name === 'AccessDeniedError'));
    const error = errors.find(error => error instanceof Error && error.name === 'AccessDeniedError');
    const hasChanged = field.hasChanged({
      [field.path]: savedValue
    }, {
      [field.path]: value
    });
    return core.jsx(fields.FieldContainer, null, core.jsx(fields.FieldLabel, {
      htmlFor: htmlID,
      field: field,
      errors: errors
    }), core.jsx(fields.FieldInput, null, core.jsx(input.Input, {
      autoComplete: "off",
      autoFocus: autoFocus,
      type: "url",
      value: canRead && value && value.originalUrl || '',
      placeholder: canRead ? undefined : error.message,
      onChange: this.onChange,
      id: htmlID
    })), value && value.originalUrl && hasChanged && core.jsx(PlaceholderPreview, {
      originalUrl: value.originalUrl,
      fieldPath: field.path
    }), value && value.originalUrl && !hasChanged && core.jsx(StyledPreview, {
      preview: value.preview,
      originalUrl: value.originalUrl,
      fieldPath: field.path
    }));
  }

}

exports.default = OEmbedField;
