// @keystone-alpha/fields/dist/chunk-7103e99f.cjs.js
'use strict';

function _interopDefault(ex) {
  return ex && typeof ex === 'object' && 'default' in ex ? ex['default'] : ex;
}

var _defineProperty = _interopDefault(require("@babel/runtime/helpers/defineProperty"));

var core = require('@emotion/core');

var React = require('react');

var React__default = _interopDefault(React);

var fields = require('@arch-ui/fields');

var controls = require('@arch-ui/controls');

class TextField extends React.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "onChange", event => {
      this.props.onChange(event.target.checked);
    });
  }

  render() {
    const _this$props = this.props,
          autoFocus = _this$props.autoFocus,
          field = _this$props.field,
          value = _this$props.value,
          errors = _this$props.errors;
    const checked = value || false;
    const htmlID = "ks-input-".concat(field.path);
    return core.jsx(fields.FieldContainer, null, core.jsx(fields.FieldLabel, {
      htmlFor: htmlID,
      field: field,
      errors: errors
    }), core.jsx(fields.FieldInput, {
      css: {
        height: 35
      }
    }, core.jsx(controls.CheckboxPrimitive, {
      autoFocus: autoFocus,
      checked: checked,
      onChange: this.onChange,
      id: htmlID
    })));
  }

}

exports.default = TextField;
