// @keystone-alpha/fields/dist/chunk-7d8267e7.esm.js
import _objectSpread from '@babel/runtime/helpers/esm/objectSpread';
import _objectWithoutProperties from '@babel/runtime/helpers/esm/objectWithoutProperties';
import _defineProperty from '@babel/runtime/helpers/esm/defineProperty';
import { jsx } from '@emotion/core';
import { Component, Fragment } from 'react';
import _extends from '@babel/runtime/helpers/esm/extends';
import Select from '@arch-ui/select';
import { Options, OptionPrimitive, CheckMark } from '@arch-ui/options';
import { RadioGroup, Radio } from '@arch-ui/filters';
import { gridSize } from '@arch-ui/theme';

const EventCatcher = props => jsx("div", _extends({
  onClick: e => {
    e.preventDefault();
    e.stopPropagation();
  }
}, props));

const SelectWrapper = props => jsx("div", _extends({
  css: {
    marginTop: gridSize * 2
  }
}, props));

class SelectFilterView extends Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      inverted: this.props.value.inverted
    });

    _defineProperty(this, "handleRadioChange", value => {
      const _this$props = this.props,
            onChange = _this$props.onChange,
            oldValue = _this$props.value;
      const inverted = value === 'does_match' ? false : true;
      onChange(_objectSpread({}, oldValue, {
        inverted
      }));
    });

    _defineProperty(this, "handleSelectChange", value => {
      const _this$props2 = this.props,
            onChange = _this$props2.onChange,
            oldValue = _this$props2.value;
      const options = [].concat(value); // ensure consistent data shape

      onChange(_objectSpread({}, oldValue, {
        options
      }));
    });
  }

  render() {
    const _this$props3 = this.props,
          innerRef = _this$props3.innerRef,
          field = _this$props3.field,
          value = _this$props3.value;
    const radioValue = value.inverted ? 'does_not_match' : 'does_match';
    const selectProps = {
      components: {
        Option: CheckMarkOption
      },
      innerRef: innerRef,
      onChange: this.handleSelectChange,
      options: field.options,
      placeholder: 'Select...',
      value: value.options
    };
    return jsx(Fragment, null, jsx(RadioGroup, {
      onChange: this.handleRadioChange,
      value: radioValue
    }, jsx(Radio, {
      value: "does_match"
    }, "Matches"), jsx(Radio, {
      value: "does_not_match"
    }, "Does not match")), jsx(SelectWrapper, null, field.options.length > 8 ? jsx(EventCatcher, null, jsx(Select, _extends({
      menuPortalTarget: document.body
    }, selectProps))) : jsx(Options, _extends({
      displaySearch: false
    }, selectProps))));
  }

}

const CheckMarkOption = _ref => {
  let children = _ref.children,
      props = _objectWithoutProperties(_ref, ["children"]);

  return jsx(OptionPrimitive, props, jsx("span", null, children), jsx(CheckMark, {
    isFocused: props.isFocused,
    isSelected: props.isSelected
  }));
};

export default SelectFilterView;
