// @keystone-alpha/fields/dist/chunk-0fcf6622.esm.js
import _defineProperty from '@babel/runtime/helpers/esm/defineProperty';
import { jsx } from '@emotion/core';
import { Component } from 'react';
import { FieldContainer, FieldLabel, FieldInput } from '@arch-ui/fields';
import { CheckboxPrimitive } from '@arch-ui/controls';

class TextField extends Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "onChange", event => {
      this.props.onChange(event.target.checked);
    });
  }

  render() {
    const _this$props = this.props,
          autoFocus = _this$props.autoFocus,
          field = _this$props.field,
          value = _this$props.value,
          errors = _this$props.errors;
    const checked = value || false;
    const htmlID = "ks-input-".concat(field.path);
    return jsx(FieldContainer, null, jsx(FieldLabel, {
      htmlFor: htmlID,
      field: field,
      errors: errors
    }), jsx(FieldInput, {
      css: {
        height: 35
      }
    }, jsx(CheckboxPrimitive, {
      autoFocus: autoFocus,
      checked: checked,
      onChange: this.onChange,
      id: htmlID
    })));
  }

}

export default TextField;
