// @keystone-alpha/fields/dist/chunk-2dc2cb1a.cjs.js
'use strict';

function _interopDefault(ex) {
  return ex && typeof ex === 'object' && 'default' in ex ? ex['default'] : ex;
}

var _objectSpread = _interopDefault(require("@babel/runtime/helpers/objectSpread"));

var luxon = require('luxon');

var dateFns = require('date-fns');

var React = require('react');

var React__default = _interopDefault(React);

var dayPicker = require('@arch-ui/day-picker');

function parseDate(value) {
  // why do we use luxon instead of date-fns/native Dates?
  // native Dates don't allow you to set custom timezones
  // yes, you can parse other timezones but they're converted
  // to the local time zone
  const dt = luxon.DateTime.fromISO(value, {
    setZone: true
  });
  return {
    date: dt.toFormat('yyyy-LL-dd'),
    time: dt.toFormat('HH:mm:ss.SSS'),
    offset: dt.toFormat('ZZ')
  };
}

function stringifyDate(date) {
  return "".concat(date.date, "T").concat(date.time).concat(date.offset);
}

const CalendarDayFilterView = props => {
  const parsedDate = props.value ? parseDate(props.value) : parseDate(new Date().toISOString());

  let handleDayChange = day => {
    props.onChange(stringifyDate(_objectSpread({}, parsedDate, {
      date: dateFns.format(day, 'YYYY-MM-DD')
    })));
  };

  let handleTimeChange = event => {
    props.onChange(stringifyDate(_objectSpread({}, parsedDate, {
      time: event.target.value
    })));
  };

  let handleOffsetChange = offset => {
    props.onChange(stringifyDate(_objectSpread({}, parsedDate, {
      offset
    })));
  };

  if (!props.filter) return null;
  const _props$field$config = props.field.config,
        yearRangeFrom = _props$field$config.yearRangeFrom,
        yearRangeTo = _props$field$config.yearRangeTo,
        yearPickerType = _props$field$config.yearPickerType;
  return React__default.createElement(dayPicker.DayTimePicker, _objectSpread({}, parsedDate, {
    htmlID: 'calendar-day-filter',
    handleDayChange,
    handleTimeChange,
    handleOffsetChange,
    yearRangeFrom,
    yearRangeTo,
    yearPickerType
  }));
};

exports.default = CalendarDayFilterView;
