// @keystone-alpha/fields/dist/chunk-1df37047.esm.js
import _objectSpread from '@babel/runtime/helpers/esm/objectSpread';
import '@babel/runtime/helpers/esm/objectWithoutProperties';
import _slicedToArray from '@babel/runtime/helpers/esm/slicedToArray';
import pluralize from 'pluralize';
import { jsx } from '@emotion/core';
import { createContext, useState, Suspense, Fragment, useContext } from 'react';
import '@babel/runtime/helpers/esm/extends';
import { Button } from '@arch-ui/button';
import { a as PreviewPlaceholder } from './chunk-30394a4e.esm.js';
let type = 'oEmbed'; // TODO: Receive this value from the server somehow. 'pluralize' is a fairly
// large lib.

const path = pluralize.plural(type);
let Context = createContext(null);
let Provider = Context.Provider;

const Embed = _ref => {
  let url = _ref.url,
      oembedData = _ref.oembedData;
  let options = useContext(Context);

  if (options.previewComponent) {
    // The adapter should implement this option
    const _options$adminMeta$re = options.adminMeta.readViews([options.previewComponent]),
          _options$adminMeta$re2 = _slicedToArray(_options$adminMeta$re, 1),
          Preview = _options$adminMeta$re2[0];

    return jsx(Preview, {
      url: url,
      options: options
    });
  } else {
    // This is a fallback so we can at least try to render _something_
    return jsx(PreviewPlaceholder, {
      data: oembedData,
      originalUrl: url
    });
  }
};

let Block = _ref2 => {
  let url = _ref2.url,
      oembedData = _ref2.oembedData,
      onChange = _ref2.onChange,
      onRemove = _ref2.onRemove;

  let _useState = useState(url),
      _useState2 = _slicedToArray(_useState, 2),
      currentValue = _useState2[0],
      setCurrentValue = _useState2[1];

  let embed = null;

  if (url) {
    embed = jsx(Suspense, {
      fallback: jsx("div", null, "Generating Preview...")
    }, jsx(Embed, {
      url: url,
      oembedData: oembedData
    }));
  }

  return jsx(Fragment, null, jsx("form", {
    onSubmit: e => {
      e.preventDefault();
      onChange(currentValue);
    }
  }, jsx("div", {
    css: {
      width: '100%',
      display: 'flex'
    }
  }, jsx("input", {
    type: "url",
    placeholder: "Enter a URL and press enter to add an embed",
    css: {
      flex: 10,
      display: 'inline',
      border: 'none',
      backgroundColor: 'transparent',
      outline: 'none',
      paddingTop: 8,
      paddingBottom: 8,
      fontSize: 18
    },
    onClick: e => {
      e.stopPropagation();
    },
    value: currentValue,
    onChange: e => {
      setCurrentValue(e.target.value);
    }
  }), jsx("div", {
    css: {
      display: 'inline-flex',
      alignItems: 'center',
      padding: 8
    }
  }, jsx(Button, {
    appearance: "danger",
    onClick: onRemove,
    type: "button"
  }, "Remove")))), embed);
};

function Sidebar(_ref3) {
  let editor = _ref3.editor;
  return jsx("button", {
    type: "button",
    onClick: () => {
      editor.insertBlock({
        type
      });
    }
  }, "Embed");
}

function Node(_ref4) {
  let node = _ref4.node,
      editor = _ref4.editor;
  return jsx(Block, {
    url: node.data.get('url'),
    oembedData: node.data.get('oembedData'),
    onRemove: () => {
      editor.removeNodeByKey(node.key);
    },
    onChange: url => {
      editor.setNodeByKey(node.key, {
        data: node.data.set('url', url)
      });
    }
  });
}

let getSchema = () => ({
  isVoid: true
});

function serialize(_ref5) {
  let node = _ref5.node;
  const url = node.data.get('url');
  const joinIds = node.data.get('_joinIds');
  const mutations = joinIds && joinIds.length ? {
    connect: {
      id: joinIds[0]
    }
  } : {
    create: {
      embed: url
    }
  };
  return {
    mutations,
    node: _objectSpread({}, node.toJSON(), {
      // Zero out the data so we don't unnecesarily duplicate the url
      data: {}
    })
  };
}

function deserialize(_ref6) {
  let node = _ref6.node,
      joins = _ref6.joins;

  if (!joins || !joins.length) {
    console.error('No embed data received when rehydrating oEmbed block');
    return;
  } // Inject the original url back into the block


  return node.set('data', node.data.set('url', joins[0].embed.originalUrl).set('oembedData', joins[0].embed));
}

export { type, path, Provider, Sidebar, Node, getSchema, serialize, deserialize };
